### PR TITLE
Mending enchantment

### LINKED ISSUE #567: Implement mending

### PR DESCRIPTION
**Pull request by Prim69**
Resolves #567.

https://user-images.githubusercontent.com/13991048/178198032-8cc3848a-1946-4d2c-87a5-54a89c3bab9d.mov

### CODE REVIEW

File: server/item/enchantment/mending.go [hunk 1]
```diff
@@ -0,0 +1,25 @@
+package enchantment
+
+import (
+	"github.com/df-mc/dragonfly/server/item"
+)
+
+// Mending is an enchantment that repairs the item when experience orbs are collected.
+type Mending struct{}
```
Comment (JustTalDevelops) [inline server/item/enchantment/mending.go:8]: Make sure to implement the Rarity function, introduced with anvils.

Comment (JustTalDevelops) [reply server/item/enchantment/mending.go:8]: Along with the `CompatibleWithEnchantment` function, and the rename from `CompatibleWith` to `CompatibleWithItem`.

File: server/item/enchantment/mending.go [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+}
+
+// Rarity ...
+func (Mending) Rarity() item.EnchantmentRarity {
+	return item.EnchantmentRarityRare
+}
+
+// CompatibleWithEnchantment ...
+func (Mending) CompatibleWithEnchantment(item.EnchantmentType) bool {
+	//_, infinity := t.(Infinity) todo: infinity
```
Comment (Sandertv) [inline server/item/enchantment/mending.go:28]: Could you uppercase this `TODO` for consistency? 😄

File: server/player/player.go [hunk 3]
```diff
@@ -2028,8 +2028,49 @@ func (p *Player) CollectExperience(value int) bool {
...
+	}
+	for _, i := range p.Armour().Items() {
+		if i.Durability() == i.MaxDurability() {
+			continue
+		}
+		if _, ok := i.Enchantment(enchantment.Mending{}); ok {
+			mendingItems = append(mendingItems, i)
+		}
+	}
+	if length := len(mendingItems); length > 0 {
```
Comment (Sandertv) [inline server/player/player.go:2048]: I'd invert this check and return if length = 0 (after moving this into its separate function.

File: server/player/player.go [hunk 4]
```diff
@@ -2028,8 +2028,49 @@ func (p *Player) CollectExperience(value int) bool {
...
+			}
+			if offHand.Equal(foundItem) {
+				p.SetHeldItems(held, repairedItem)
+			} else if held.Equal(foundItem) {
+				p.SetHeldItems(repairedItem, offHand)
+			} else if slot, ok := p.Armour().Inventory().First(foundItem); ok {
+				_ = p.Armour().Inventory().SetItem(slot, repairedItem)
+			}
+		}
+	}
```
Comment (Sandertv) [inline server/player/player.go:2065]: I think it'd be a good idea to move this entire block into its own function. We'd want that to return the amount of leftover XP so we can use that for the rest of the code after it in this function.

**Review by Sandertv — CHANGES_REQUESTED** [review]
I've left a couple of comments. Functionally the code looks fine to me, so it shouldn't be too much more work.

File: server/player/player.go [hunk 5]
```diff
@@ -2028,8 +2028,54 @@ func (p *Player) CollectExperience(value int) bool {
...
+	if value > 0 {
+		return p.AddExperience(value) > 0
+	}
+
+	p.PlaySound(sound.Experience{})
+	return true
+}
+
+// handleMending handles the mending enchantment when collecting experience, it then returns the leftover experience.
+func (p *Player) handleMending(value int) int {
```
Comment (Sandertv) [inline server/player/player.go:2042]: Would probably name this method `mendItems(xp int)` to be a bit more descriptive.

**Review by Sandertv — COMMENTED** [review]
Just one last comment, feel free to ping me on Discord once you've resolved it. Should be good to merge after it.

**Review by JustTalDevelops — APPROVED** [review]

**Pull request merged**